### ISSUE TITLE
Implement local functions in GOOL

### ISSUE THREAD

**Issue by bmaclach** [JacquesCarette/Drasil#1987]
The ODE solvers for Python and C# use local functions in some form (nested function for Python, anonymous function for C#). 

Nested functions are only available in Python and C#. Anonymous functions are available in all 4 target languages, so it makes sense to implement them in GOOL. For Python and C#, we could even set it up so that if a user tries to assign a local function to a variable, the renderer generates a nested function instead. This would require a new type for local state for `Value`s, but that will probably be useful in many other cases too.

**Comment by JacquesCarette** [comment]
I think it would make sense to elaborate on a potential design before implementing this. Conceptually, "local functions" should not really be different than a function-valued local variable. Of course, different languages implement this different [and sometimes not at all].  But we can always generate (hopefully inaccessbile) global functions with long names to get around that.

**Comment by bmaclach** [comment]
@JacquesCarette So I had already started working on this, but what I've done seems to line up with what you're expecting.

I agree that a local function is just a function-valued local variable. What I've worked on so far is adding a `lambda` function to GOOL, which can be used to create (representations of) values with function type.

What I was planning on doing next was using `State` to detect when a user tries to assign a lambda to a variable in Python and C#, because those 2 languages support nested, non-anonymous local functions, and so we should do that instead of assigning an anonymous function a variable name. 

But based on what you said in #1991 about avoiding code analysis wherever possible, maybe I should instead implement a separate function for the case where a local function is assigned a variable name. Something like `funcDecDef` (analogous to existing functions like `varDecDef`)? But then we are trusting the user not to use this inappropriately, whereas with code analysis the responsibility is in GOOL's hands, so we would only be doing this where it makes sense.
